Feature/faiss integration

Jerry01020 opened the pull request:
在 SageFlow 中集成了 FAISS 作为向量相似度连接的后端，支持 IVF 和 HNSW 算法。同时，针对多算子并发场景下的混合并行（Hybrid Parallelism）问题进行了优化，通过自动管理 OpenMP 线程状态，避免了严重的线程争用，显著提升了系统稳定性与吞吐量。
主要变更
算法集成：新增 [FAISS_IVF]和 [FAISS_HNSW]算法支持。
并行优化：在 JoinStrategyFactory 中增加智能检测逻辑。当算子并行度 > 1 时，自动禁用 FAISS 内部的 OpenMP 多线程，防止 CPU 资源过载。
基准测试：新增 [run_faiss_omp_benchmark.py]脚本及相关测试配置，验证了 IVF 算法在开启 OMP 时有约 16% 的性能提升（单算子场景）。
文档更新：新增了 FAISS 集成总结文档。

Jerry01020 committed:
初步完成HDR+Tree

Jerry01020 committed:
添加中文注释

Jerry01020 committed:
实现集成HDR Tree和简单剪枝逻辑

Jerry01020 committed:
完成PCA训练和RTree构建

Jerry01020 committed:
实现森林级剪枝

Jerry01020 committed:
基本完成逻辑功能

Jerry01020 committed:
完成HDRTree基本逻辑功能

Jerry01020 committed:
优化单元测试

Jerry01020 committed:
优化HDR_Tree效率

Jerry01020 committed:
Merge branch 'main-dev' of https://github.com/intellistream/sageFlow into main-dev

Jerry01020 committed:
优化HDR_Tree索引

Jerry01020 committed:
能通过join_datasource_modes测试

Jerry01020 committed:
HDR_Tree SIFT数据集加载测试和随机数据内存生成测试

Jerry01020 committed:
可通过现有完整流程测试的HDR_Tree

Jerry01020 committed:
add SUMMARY_OF_HDR_Tree

Jerry01020 committed:
add SUMMARY_OF_HDR_Tree

Jerry01020 committed:
修复HDR_Forest测试不通过的问题

Jerry01020 committed:
调整测试内容

Jerry01020 committed:
refactor(index): optimize logging and error handling in HDR implementation

- Replace std::cerr with SAGEFLOW_LOG_ERROR in hdr_tree.cpp for better integration with project logging
- Remove unnecessary debug output in hdr_tree.cpp
- Replace iostream with logger.h in hdr_forest.cpp
- Add proper error logging for missing records and storage manager in hdr_forest.cpp

Jerry01020 committed:
feat: Implement thread-safe FaissIndex with auto-training and Chinese comments

Jerry01020 committed:
Merge branch 'main-dev' into feature/faiss-integration

Jerry01020 committed:
Fix FAISS integration after merge: update config, factory, and tests

Jerry01020 committed:
Fix FAISS integration after merge: update config, factory, and tests

Jerry01020 committed:
feat: Add configuration option to disable OpenMP in FAISS index

- Added faiss_disable_omp to JoinStrategyConfig
- Updated FaissIndex to accept disable_omp parameter
- Updated ConcurrencyManager and JoinStrategyFactory to propagate the config
- Default behavior is to disable OpenMP to avoid thread oversubscription

Jerry01020 committed:
feat: Add FAISS OMP benchmark and optimize hybrid parallelism handling

- Added scripts/run_faiss_omp_benchmark.py for performance testing
- Updated JoinStrategyFactory to automatically disable FAISS OMP when operator parallelism > 1 to prevent thread contention
- Updated integration test cases and baseline tests

Jerry01020 committed:
feat: Add FAISS OMP benchmark and optimize hybrid parallelism handling

- Added scripts/run_faiss_omp_benchmark.py for performance testing
- Updated JoinStrategyFactory to automatically disable FAISS OMP when operator parallelism > 1 to prevent thread contention
- Updated integration test cases and baseline tests

Jerry01020 committed:
add FAISS_INTEGRATION_SUMMARY

Jerry01020 committed:
Fix: Force Generic BLAS to resolve FAISS build failure on environments with incomplete MKL

Jerry01020 committed:
Fix: Use pkg-config for BLAS detection to support CI environments

Jerry01020 committed:
ci: fix missing BLAS dependencies in performance test workflow

ZeroJustMe commented on src/index/faiss_index.cpp:
```diff
@@ -0,0 +1,201 @@
+  std::shared_lock<std::shared_mutex> lock(mutex_);
+  
+  try {
+    faiss_index_->range_search(1, reinterpret_cast<const float*>(record.data_.data_.get()), join_similarity_threshold, &res);
```
这里直接用join_similarity_threshold来作为参数应该有问题？range_search的参数是radius半径，我们的阈值是用ComputeEngine::Similarity的指数函数算出来的，这里需要换算一下

Jerry01020 replied:
后续版本调整这里的处理逻辑

ZeroJustMe commented on scripts/run_integration_test.py:
```diff
@@ -46,6 +46,7 @@
     'clustered_join': '*clustered_join*',
     's3j': '*s3j*',
     'vsjoin': '*vsjoin*',
+    'faiss': '*faiss*',
```
这里不用区分faiss hnsw和ivf吗

Jerry01020 replied:
这应该是最初版本的逻辑，只注重观察faiss的测试效果所以没有进行划分，后续视情况做出调整

ZeroJustMe commented on src/operator/utils/join_strategy_factory.cpp:
```diff
@@ -122,6 +138,12 @@ std::unique_ptr<BaseMethod> JoinStrategyFactory::createJoinMethod(
         case JoinAlgorithm::IVF:
             return createIvfMethod(config, concurrency_manager, 
                                    left_index_id, right_index_id);
+        case JoinAlgorithm::FAISS_IVF:
+            return createIvfMethod(config, concurrency_manager, 
+                                   left_index_id, right_index_id);
+        case JoinAlgorithm::FAISS_HNSW:
+            return createHnswMethod(config, concurrency_manager, 
+                                    left_index_id, right_index_id);
         case JoinAlgorithm::HNSW:
             return createHnswMethod(config, concurrency_manager, 
                                     left_index_id, right_index_id);
```
这里faiss版本的和之前的索引是用的同一个方法吗，区别只是cc_manager里的索引不同？

Jerry01020 replied:
其他的处理逻辑基本是一样的

ZeroJustMe commented on scripts/run_integration_test.py:
```diff
@@ -46,6 +46,7 @@
     'clustered_join': '*clustered_join*',
     's3j': '*s3j*',
     'vsjoin': '*vsjoin*',
+    'faiss': '*faiss*',
```
这里可以区分一下faiss_ivf和hnsw？

Jerry01020 replied:
后续版本准备优化这里的分类

ZeroJustMe commented on scripts/run_integration_test.py:
用集成测试脚本跑一下这两个新加的方法的recall和耗时的指标，最好能有一个和之前版本的ivf和hnsw的性能对比

Jerry01020 replied:
后续版本添加这个对比

ZeroJustMe reviewed (commented):
join_operator里面没有对新加的索引进行创建和参数初始化吗，现在还没有在整个pipeline里面测一遍这两个方法吗？还需要补充一下faiss ivf和hnsw整体的recall和延迟指标

ZeroJustMe committed:
feat(join): QIQ concurrency optimization and project refactoring (#89)

* refactor: restructure project layout and modularize VSJoin components

* feat(join): add QIQ phase metrics and re-enable Q2 in high-parallelism path

* test(perf): add QIQ stats output and enable HDR-Tree in performance test

* docs: add QIQ concurrency optimization plan and analysis

* chore: add TODO issue links via todo-to-issue-action

* fix integration test bug

* fix(ci): remove commits and DIFF_URL params to fix todo-to-issue-action

The todo-to-issue-action was failing because:
- DIFF_URL: '' caused LocalClient mode to be used
- LocalClient runs inside Docker container which lacks git

Solution: Remove commits and DIFF_URL parameters, let the action
automatically use GitHubClient mode for pull_request events.
GitHubClient fetches diff via GitHub API, no local git needed.

---------

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

ZeroJustMe committed:
feat(join): QIQ concurrency optimization and project refactoring (#89)

* refactor: restructure project layout and modularize VSJoin components

* feat(join): add QIQ phase metrics and re-enable Q2 in high-parallelism path

* test(perf): add QIQ stats output and enable HDR-Tree in performance test

* docs: add QIQ concurrency optimization plan and analysis

* chore: add TODO issue links via todo-to-issue-action

* fix integration test bug

* fix(ci): remove commits and DIFF_URL params to fix todo-to-issue-action

The todo-to-issue-action was failing because:
- DIFF_URL: '' caused LocalClient mode to be used
- LocalClient runs inside Docker container which lacks git

Solution: Remove commits and DIFF_URL parameters, let the action
automatically use GitHubClient mode for pull_request events.
GitHubClient fetches diff via GitHub API, no local git needed.

---------

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

ZeroJustMe commented:
新合进去一个pr，记得先rebase一下最新的main-dev解一下冲突再进行修改

Jerry01020 committed:
Merge branch 'main-dev' into feature/faiss-integration

Jerry01020 committed:
fix(index): correct threshold conversion in FaissIndex::query_for_join

- Implemented inverse logic for ComputeEngine::Similarity (exponential decay)
- Converted similarity threshold to correct Faiss search radius (L2 squared)
- Added specific handling for Inner Product metric

Jerry01020 committed:
Fix: Complete Faiss integration with correct metric logic and HNSW fallback

- Fix(FaissIndex): Initialize dimension_ correctly to prevent mismatch errors.
- Feat(FaissIndex): Implement KNN fallback for HNSW range search (Task 3).
- Fix(Math): Correct L2/IP threshold conversion formulas (Task 1).
- Test: Update deterministic unit tests and enable integration tests.
- Tools: Enhance test scripts to support specific faiss_ivf/hnsw modes.

Jerry01020 committed:
WIP: Refactor Faiss logic with HNSW fallback (Integration tests tuning)

- Fix(FaissIndex): Correct dimension initialization to prevent mismatches.
- Feat(FaissIndex): Implement KNN fallback strategy for HNSW range search.
- Fix(Math): Update metric conversion logic for L2/IP thresholds.
- Test: Integration tests are currently unstable due to insufficient data size (tuning in progress).

Jerry01020 committed:
implement faiss
